### PR TITLE
Add *good* error messages for extra fields on resources. 

### LINKED ISSUE #197: Improve error message when user forgets "properties"?
Is it possible to improve the error message if someone places a YAML attribute that should be under properties? (This idea occurred to me because the error message for specifying an invalid property name is so helpful and the onboarding was so gentle for me and because I made this mistake several times while writing my first Pulumi YAML programs.)

e.g.:

```yaml
myBucket:
  type: aws:s3:bucket
  bucket: 'my-bucket-name'
```

Could result in some error message like:

```
Resource 'myBucket' has invalid key 'bucket'. Did you mean to put this key for the resource under 'properties', e.g.:

myBucket:
  # ...
  properties:
    bucket: 'my-bucket-name'
```

### PR DESCRIPTION
**Pull request by iwahbe**
Fixes #197

Since this adds a hard error, this change is a breaking change. We are in since we haven't hit 1.0, I think this is ok.

The generated error looks like:
```
❯ pulumi preview
Previewing update (dev)

View Live: https://app.pulumi.com/iwahbe/yaml-example/dev/previews/7e608ff4-33dc-41b8-ad8e-c8561a4198a4

     Type                 Name              Plan       Info
 +   pulumi:pulumi:Stack  yaml-example-dev  create     2 errors; 8 messages
 
Diagnostics:
  pulumi:pulumi:Stack (yaml-example-dev):
    error: Error: Resource aws:s3/bucket:Bucket has invalid key 'bucket'
    
      on Pulumi.yaml line 8:
       8:     bucket: "my-bucket-name"
    
    'bucket' exists under properties, e.g.
    
    myBucket:
      # ...
      properties:
        bucket: my-bucket-name
    error: an unhandled error occurred: failed to evaluate template
 
    Error: Resource aws:s3/bucket:Bucket has invalid key 'bucket'
      on Pulumi.yaml line 8:
       8:     bucket: "my-bucket-name"
    'bucket' exists under properties, e.g.
    myBucket:
      # ...
      properties:
        bucket: my-bucket-name
```

If there isn't an exact match, we provide up to 5 suggestions. We ensure that all suggestions are within a certain (3) edit distance. This is a hyper parameter we should experiment with tuning.

```
X pulumi preview
Previewing update (dev)

View Live: https://app.pulumi.com/iwahbe/yaml-example/dev/previews/37736e47-5651-4aa1-a009-bbc3e3339424

     Type                 Name              Plan       Info
 +   pulumi:pulumi:Stack  yaml-example-dev  create     2 errors; 6 messages
 
Diagnostics:
  pulumi:pulumi:Stack (yaml-example-dev):
    error: Error: Resource aws:s3/bucket:Bucket has invalid key 'version'
    
      on Pulumi.yaml line 8:
       8:     version: 1.2.3
    
    Did you mean any of:
      - Version under options
      - versioning under properties
    error: an unhandled error occurred: failed to evaluate template
 
    Error: Resource aws:s3/bucket:Bucket has invalid key 'version'
      on Pulumi.yaml line 8:
       8:     version: 1.2.3
    Did you mean any of:
      - Version under options
      - versioning under properties
 ```

### CODE REVIEW

**Comment by AaronFriel** [comment]
@jkodroff I know you like to see these error messages improving

**Comment by jkodroff** [comment]
You're darn tootin' I do. This is definitely better!

A few suggestions on the wording (the general idea being to make it more unambiguous):

1. Suggest always quoting a key or anything the user supplied for clarity, e.g. ''`'properties'`
2. `Did you mean any of:` -> `You may have meant one of the following:`
3. Indent and number the corrections if possible without undue effort, e.g.:

    ```yaml
    1.  properties:
          versioning:
    ```
4. Capitalization of Error/error should be consistent if feasible.
5. Make blank lines consistent if feasible.

Would also love to hear @susanev or a teammate's suggestions as they have expertise beyond mine.

File: pkg/pulumiyaml/analyser.go [hunk 1]
```diff
@@ -49,14 +51,63 @@ func (tc *typeCache) typeResource(r *runner, node resourceNode) bool {
...
+					{Name: "options", Properties: allOptions},
+					{Name: k, Properties: valid},
+				},
+				DistanceLimit: 3,
+			}
+			for i := 0; i < o.Len(); i++ {
+				prop := o.Index(i)
+				key := prop.Key.Value()
+				keyLower := strings.ToLower(key)
+				isValid := false
```
Comment (jkodroff) [inline pkg/pulumiyaml/analyser.go:84]: Can just be `valid` or if we change `valid` above to something more explicit, e.g. `validKeys`.

**Comment by iwahbe** [comment]
With regards to numbering the bullets, it gets easy to confuse with line numbers:
(real example)
```
    Error: Resource aws:s3/bucket:Bucket has invalid key 'version'
      on Pulumi.yaml line 3:
       3:     version: 1.2.3
    You may have meant one of the following:
      1. Version under options
      2. versioning under properties
```

I changed it to lettering with open parentheses:
```
    Error: Resource aws:s3/bucket:Bucket has invalid key 'version'
      on Pulumi.yaml line 3:
       3:     version: 1.2.3
    You may have meant one of the following:
      a) Version under options
      b) versioning under properties
```

Let me know what you think.

**Review by AaronFriel — APPROVED** [review]

**Pull request merged**